ardupilot: add DO_WINCH command

rmackay9 opened the pull request:
This PR does the following:

- adds a new DO_WINCH command for use within a mission-item, command-long or command-int message to support controlling a winch for lowering packages from a vehicle
- changes the order in which the MAV_CMD_GIMBAL_RESET and MAV_CMD_SET_FACTORY_TEST_MODE commands appear in ardupilotmega.xml so that the order is consistent with the command-id value.  This is a non-functional change.

WickedShell commented on message_definitions/v1.0/ardupilotmega.xml:
```diff
@@ -210,6 +210,16 @@
+      <entry value="42600" name="MAV_CMD_DO_WINCH">
+        <description>Command to operate winch</description>
+        <param index="1">winch number (0 for the default winch, otherwise a number from 1 to max number of winches on the vehicle)</param>
+        <param index="2">action (0=disable, 1=length control, 2=rate control.  See WINCH_ACTIONS enum)</param>
```
I'd remove the 0=disable, 1=length control etc. It's not scalable, and it's why we have the comment that tells you to lookup the enum.

WickedShell replied:
Actually I take it back. I find this to be a confusing command. Why do you need the action at all? You already have full control over both length and rate in the params. If you only want to control the one why not just set the other to 0? (or NaN for no control)

rmackay9 replied:
"0" has a real meaning for both rate-control and length-control so I don't think we can use that to mean "ignore".  For example maintaining the rope at it's current length is action=length-control with length of 0.  I'm definitely not a fan of asking users to enter NaN.  I suspect many users don't know what NaN is.
I agree that we could use length-control and then use the rate field to specific the desired rate at which we get to the desired length.  I haven't implemented that for now but it could be added later.

WickedShell replied:
Length control is purely a relative length command as you've implemented it here. The proposed XML file doesn't allow the usage of both rate and length control as presented. (And if it does then why is param2 used for a enum?

rmackay9 replied:
i've added the length+rate control now.
one way or the other, we need to have an action field to specify relaxing the winch.

rmackay9 commented:
@amilcarlucas, OK, I've modified the release-length call to accept the rate as a maximum and modified the vehicle code to pass in the rate field from the mavlink command.

OXINARF commented on message_definitions/v1.0/ardupilotmega.xml:
```diff
@@ -210,6 +210,16 @@
+      <entry value="42600" name="MAV_CMD_DO_WINCH">
+        <description>Command to operate winch</description>
+        <param index="1">winch number (0 for the default winch, otherwise a number from 1 to max number of winches on the vehicle)</param>
+        <param index="2">action (0=disable, 1=length control, 2=rate control.  See WINCH_ACTIONS enum)</param>
```
Shouldn't _disable_ be *relax* instead?

rmackay9 replied:
OK, changed from "disable" to "relaxed"
